// @getmash/client-sdk/dist/rpc/RPCApi.js
import JsonRPCEngine from "@getmash/jsonrpc-engine";
import RPCMethods from "./methods.js";
import Targets from "./targets.js";
var FiatCurrency;
(function (FiatCurrency) {
    FiatCurrency["Usd"] = "USD";
})(FiatCurrency || (FiatCurrency = {}));
class MashRPCApi {
    engine;
    constructor(origin, target) {
        this.engine = new JsonRPCEngine({
            name: Targets.MashSDK,
            targetName: Targets.Wallet,
            targetWindow: target || undefined,
            targetOrigin: origin,
        });
    }
    init(earnerID, position) {
        return this.engine.call(RPCMethods.Init, { id: earnerID, position });
    }
    access(resourceID) {
        return this.engine.call(RPCMethods.Access, resourceID);
    }
    donate() {
        return this.engine.call(RPCMethods.Donate);
    }
    getAutopayAuthorization() {
        return this.engine
            .call(RPCMethods.GetAutopayAuthorization)
            .then(res => res.authorization);
    }
    getResourceCost(resourceID) {
        return this.engine
            .call(RPCMethods.GetResourceCost, resourceID)
            .then(res => res.cost);
    }
}
export default MashRPCApi;
