// @getmash/client-sdk/dist/iframe/position.js
import { WalletButtonFloatPlacement, WalletButtonFloatSide, } from "../api/routes.js";
export var FloatLocation;
(function (FloatLocation) {
    FloatLocation["BottomRight"] = "bottom-right";
    FloatLocation["BottomLeft"] = "bottom-left";
})(FloatLocation || (FloatLocation = {}));
/**
 * Validates the string value is one of the accepted FloatSides
 * @param location string
 * @returns WalletButtonFloatSide
 */
function normalizeFloatSide(location) {
    if (!location ||
        !Object.values(WalletButtonFloatSide).includes(location)) {
        return WalletButtonFloatSide.Right;
    }
    return location;
}
/**
 * Validates the string value is one of the accepted FloatPlacements
 * @param location string
 * @returns WalletButtonFloatPlacement
 */
function normalizeFloatPlacement(location) {
    if (!location ||
        !Object.values(WalletButtonFloatPlacement).includes(location)) {
        return WalletButtonFloatPlacement.Default;
    }
    return location;
}
function getDesktopLocation(desktop) {
    return {
        floatSide: normalizeFloatSide(desktop?.floatSide),
        floatPlacement: normalizeFloatPlacement(desktop?.floatPlacement),
        customShiftConfiguration: {
            horizontal: desktop?.customShiftConfiguration?.horizontal || 0,
            vertical: desktop?.customShiftConfiguration?.vertical || 0,
        },
    };
}
function getMobileLocation(mobile) {
    return {
        floatSide: normalizeFloatSide(mobile?.floatSide),
    };
}
export function getWalletPosition(desktop, mobile) {
    return {
        desktop: getDesktopLocation(desktop),
        mobile: getMobileLocation(mobile),
    };
}
export function formatPosition(position) {
    const formattedPosition = getWalletPosition();
    if (position?.desktop?.floatLocation === FloatLocation.BottomLeft) {
        formattedPosition.desktop.floatSide = WalletButtonFloatSide.Left;
    }
    else if (position?.desktop?.floatLocation === FloatLocation.BottomRight) {
        formattedPosition.desktop.floatSide = WalletButtonFloatSide.Right;
    }
    if (position?.desktop?.shiftLeft || position?.desktop?.shiftRight) {
        formattedPosition.desktop.floatPlacement =
            WalletButtonFloatPlacement.Custom;
        formattedPosition.desktop.customShiftConfiguration.horizontal =
            position.desktop.shiftLeft || 0;
        formattedPosition.desktop.customShiftConfiguration.horizontal =
            position.desktop.shiftRight || 0;
    }
    if (position?.desktop?.shiftUp) {
        formattedPosition.desktop.floatPlacement =
            WalletButtonFloatPlacement.Custom;
        formattedPosition.desktop.customShiftConfiguration.vertical =
            position.desktop.shiftUp || 0;
    }
    if (position?.mobile?.floatLocation === FloatLocation.BottomLeft) {
        formattedPosition.mobile.floatSide = WalletButtonFloatSide.Left;
    }
    else if (position?.mobile?.floatLocation === FloatLocation.BottomRight) {
        formattedPosition.mobile.floatSide = WalletButtonFloatSide.Right;
    }
    return formattedPosition;
}
