// @getmash/client-sdk/dist/widgets/widgets.js
export const DeprecatedWidgets = [
    "mash-text-reveal",
    "mash-download-btn",
    "mash-paywall",
    "mash-youtube-player",
    "mash-donate-btn",
];
export const Widgets = {
    Accordion: { path: "accordion/accordion.js", element: "mash-accordion" },
    Boost: { path: "boost/boost.js", element: "mash-boost-button" },
    ButtonRevealer: {
        path: "content/button-revealer.js",
        element: "mash-button-revealer",
    },
    ContentRevealer: {
        path: "content/content-revealer.js",
        element: "mash-content-revealer",
    },
    DonateButton: {
        path: "donate/donate-button.js",
        element: "mash-donate-button",
    },
    DownloadButton: {
        path: "download/download-button.js",
        element: "mash-download-button",
    },
    Link: { path: "link/link.js", element: "mash-link" },
    VideoPlayer: { path: "media/video-player.js", element: "mash-video-player" },
};
/**
 * Inject widgets into the page
 * @param baseURL Base URL of endpoint where widgets are hosted
 */
export function injectWidgets(baseURL) {
    const scripts = Object.values(Widgets).map(value => {
        const script = window.document.createElement("script");
        script.async = true;
        script.src = `${baseURL}/${value.path}`;
        return script;
    });
    window.document.head.append(...scripts);
}
function _isWidgetOnPage() {
    const widgets = Object.values(Widgets);
    // Check for new widgets
    for (let i = 0; i < widgets.length; i++) {
        const el = window.document.querySelector(widgets[i].element);
        if (el)
            return true;
    }
    // Check for old widgets
    for (let i = 0; i < DeprecatedWidgets.length; i++) {
        const el = window.document.querySelector(DeprecatedWidgets[i]);
        if (el)
            return true;
    }
    return false;
}
/**
 * Check if a known Mash Widget exists on the page
 * @returns boolean
 */
export function isWidgetOnPage() {
    return new Promise(resolve => {
        if (document.readyState === "loading") {
            document.addEventListener("DOMContentLoaded", () => resolve(_isWidgetOnPage()));
            return;
        }
        resolve(_isWidgetOnPage());
    });
}
